### DISCUSSION TITLE
Develop version vs Stable version

### DISCUSSION THREAD

**Discussion by zero88** [SoftEtherVPN/SoftEtherVPN#1305] [General]
Currently, I intend building IoT solution with a thousand multi-tenant IoT devices(linux armv7) under SoftEtherVPN network.
The server will be deployed on cloud: Google cloud in GCE (in docker) or GKE.
My strategy is opensource in another repository and will provide under Apache License 2.0. So please give me any copyright from current repository that I need to take care in my repository.

For my project status:

- Deploy automatically on device side is almost finished.
- Right now, I move on next step with VPN server to make VPN manager on web based that adapt with my objective.

I want to know a status of some ticket:

- #615
- https://github.com/SoftEtherVPN/SoftEtherVPN#built-in-softether-server-vpn-json-rpc-api-suite

Because no actual changelog, then my questions are:
- These tickets/features are available on https://github.com/SoftEtherVPN/SoftEtherVPN_Stable or not? If not, so I need to use release version in https://github.com/SoftEtherVPN/SoftEtherVPN.
- Any drawbacks if using develop version in production server?
- Is release tag in develop repository is reliable like 90% that compare to Stable version?
- Any matrix version compatible between vpnserver/vpnclient/vpnbridge? For example: using vpnclient-v4.34-9745-beta on stable repository with vpnserver-v5 on develop repository and otherwise

**Comment by zero88** [comment]
@davidebeatrici 
Thank you for your response. I will add this [license](https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/bin/hamcore/legal.txt) to my project.

**Comment by zero88** [comment]
For ticket #615, I don't know it is ready for stable repository or not. After an investigation, I found it doesn't have on stable repository.
(By find code `UNIX_SVC_ARG_FOREGROUND`)
Could you confirm? @davidebeatrici @chipitsine 

This is my log in latest stable branch
```
vpnserver_1  | The SoftEther VPN Server service has been started.
vpnserver_1  | 
vpnserver_1  | Let's get started by accessing to the following URL from your PC:
vpnserver_1  | 
vpnserver_1  | https://172.18.0.2:5555/
vpnserver_1  |   or
vpnserver_1  | https://172.18.0.2/
vpnserver_1  | 
vpnserver_1  | Note: IP address may vary. Specify your server's IP address.
vpnserver_1  | A TLS certificate warning will appear because the server uses self signed certificate by default. That is natural. Continue with ignoring the TLS warning.
vpnserver_1  | 
vpnserver_1  | [INFO  tini (1)] Spawned child process '/app/vpnserver/vpnserver' with pid '7'
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [DEBUG tini (1)] Received SIGCHLD
vpnserver_1  | [DEBUG tini (1)] Reaped child with pid: '7'
vpnserver_1  | [INFO  tini (1)] Main child exited normally (with status '0')
vpnserver_1  | [TRACE tini (1)] No child to reap
vpnserver_1  | [TRACE tini (1)] Exiting: child has exited
docker_vpnserver_1 exited with code 0
```

**Reply by davidebeatrici** [reply]
Confirmed.

**Comment by zero88** [comment]
And this [json-rpc-api](https://github.com/SoftEtherVPN/SoftEtherVPN#built-in-softether-server-vpn-json-rpc-api-suite), isn't available on latest stable repository, right?
Even, I add these configurations (default is `false` in latest stable):

```
bool UseKeepConnect true
bool UseWebTimePage true
bool UseWebUI true
```

**Reply by davidebeatrici** [reply]
It's available.

**Reply by zero88** [reply]
How to ensure if it is available, for quick test, man?
I'm running in docker and using `curl` to both port 443/8443 and 5555/5555 inside and outside docker container, still unresponsive.

**Reply by davidebeatrici** [reply]
https://<vpn_server_hostname>:<port>/admin/

**Reply by zero88** [reply]
mean without port?
The last attempt, I already connected to server jsonrpc using `POST` method (first time I use `GET` :smile:)
However, still not open admin page, might be need to add authentication header, coz server drop connection immediately.
Anyway, I just use `server jsonrpc`, so will check it later

**Comment by zero88** [comment]
Should be `-DSKIP_CPU_FEATURES=1` in [BUILD_UNIX](https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/BUILD_UNIX.md#build-without-cpu_features)

Should be `make -C tmp` in [BUILD_UNIX](https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/BUILD_UNIX.md#build-from-source-code-and-install)
```bash
-- Generating done
-- Build files have been written to: /app/vpnserver/tmp

The Makefile is generated. Run 'make -C tmp' to build SoftEther VPN.
```

**Reply by metalefty** [reply]
What does "should" mean here?

**Reply by metalefty** [reply]
> The Makefile is generated. Run 'make -C tmp' to build SoftEther VPN.

Where does this come from? Which are you talking about Stable or Developer version?

**Reply by zero88** [reply]
develop version. I tried building follow BUILD_UNIX, these are some correct steps that I need to update to make the build is success.
I'm not sure g++/make version from yours, or any implicit from your env, then I wrote "should"

**Reply by zero88** [reply]
More clarify, in document, it is `make -C build` but mine is `make -C tmp`

**Reply by metalefty** [reply]
It has already migrated from `make -C tmp` to `make -C build` We don't to need to bring it back to `tmp` version.

```
$ git describe --tags
5.01.9674-284-gb7e7d3b0
$ ./configure
(snip)
-- Build files have been written to: /tmp/SoftEtherVPN/build

The Makefile is generated. Run 'make -C build' to build SoftEther VPN.

**Reply by zero88** [reply]
@metalefty This is mine from fresh installation in [docker](https://github.com/zero88/qwe-vpn/blob/main/docker/dockerfile/sevpn.src.debian.Dockerfile#L17-L19) in latest release [5.01.9674](https://github.com/SoftEtherVPN/SoftEtherVPN/releases/tag/5.01.9674)

```
SoftEtherVPN-SoftEtherVPN-039cd8e/systemd/softether-vpnserver.service
---------------------------------------------------------------------
SoftEther VPN for Unix

Copyright (c) all contributors on SoftEther VPN project in GitHub.
Copyright (c) Daiyuu Nobori, SoftEther Project at University of Tsukuba, and SoftEther Corporation.

Licensed under the Apache License, Version 2.0 (the License).

Read and understand README, LICENSE and WARNING before use.
---------------------------------------------------------------------

Welcome to the corner-cutting configure script !

'rpm' executable not found, CPack will generate DEB packages.

-- The C compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for sys/auxv.h
-- Looking for sys/auxv.h - found
-- Build date: 23/03/2021
-- Build time: 09:33:57
-- Found Curses: /usr/lib/x86_64-linux-gnu/libcurses.so
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.1.1d")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
-- Configuring done
-- Generating done
-- Build files have been written to: /app/vpnserver/tmp

The Makefile is generated. Run 'make -C tmp' to build SoftEther VPN.
make: Entering directory '/app/vpnserver/tmp'
```

**Reply by chipitsine** [reply]
I wonder if 5.01.9674 is released with "build" or "tmp"

**Reply by metalefty** [reply]
@zero88 Indeed. The document is wrong as of 5.01.9674  but it is already fixed after the release. So you just need to refer the latest document.  I didn't know you're talking about 5.01.9674 until you explicitly mention it. So there was a discrepancy because I was talking about 5.01.9674-284-gb7e7d3b0 (master). Now it's clear.

**Reply by zero88** [reply]
@metalefty ya, my mistake. Normally, I always build on top of release version

**Comment by zero88** [comment]
One more question(sorry I'm not C guy), after `cmakes`, what binary/resource that I need to zip and delivery artifact.
Tons of things after build, `145 directories, 2167 files` just inside `vpnserver`
@davidebeatrici

**Reply by davidebeatrici** [reply]
`vpnserver`, `libcedar.so`, `libmayaqua.so` and `hamcore.se2`.

**Reply by zero88** [reply]
Thanks

**Reply by davidebeatrici** [reply]
No problem.

**Comment by zero88** [comment]
For this one:
> Any matrix version compatible between vpnserver/vpnclient/vpnbridge? For example: using vpnclient-v4.34-9745-beta on stable repository with vpnserver-v5 on develop repository and otherwise

As @davidebeatrici confirmed in github issue, I can understand:
- New server version can handle connection from lower version vpnclient. `OK`
- Lower server version can handle connection from newer version vpnclient. Still `OK` (in theory)?

**Comment by zero88** [comment]
Currently, in stable latest version, seems it doesn't support `ARM64`? Do you have plan for support it?
One curious, which your build system for multiple arch? 
Currently, I'm building with docker buildx with multiple arch. I can contribute a workflow to build artifact once I finish my task (if you want)
